src/Elastic.Apm/Helpers/StacktraceHelper.cs

```diff
@@ -28,14 +28,15 @@ internal static List<CapturedStackFrame> GenerateApmStackTrace(StackFrame[] fram
 			try
 			{
 				retVal.AddRange(from item in frames
+					let fileName = item?.GetMethod()
+						?.ReflectedType?.FullName //see: https://github.com/elastic/apm-agent-dotnet/pull/240#discussion_r289619196
```

Comment:
I thought by using `DeclaringType` it's not possible to unambiguously what code is executed. 
From the stackoverflow example, with that we would show `Base in Method`, but that's not what we want... the code executed is acutally in `Derived in Method`. 

Or am I wrong? 🤔

Reply:
Well, it depends on what you are going to do with class name. Let's say you have `Base` class with `SomeMethod()` method that throws an exception and `Derived` class that inherits from `Base`. I would argue seeing `Base` at the stack trace is more useful even if the instance for which `SomeMethod()` was called is `Derived`. It is because `Base` is where you would be able to see source code for `SomeMethod()` and `Derived` class won't have.

Reply:
Of course, if  `SomeMethod()` is `virtual` and `Derived` overrides it then we should see `Derived` and it will be as expected if we use `DeclaringType`.

Reply:
I get your example. 

But what if `Derived` overrides `SomeMethod`? In that case showing `Base` seems to be wrong to me, while in your example showing `Derived` isn't completely wrong - you can end up in the right place of the code simply looking for the method starting from `Derived` going through parent classes as long as you find the method.

Do we know what other tools do?

Reply:
> Of course, if SomeMethod() is virtual and Derived overrides it then we should see Derived and it will be as expected if we use DeclaringType.

If that's the case and there aren't any issues with that I'm ok with that.

Reply:
Just to be on the safe side I also tested `DeclaringType` with `interface` and implementing `class` - `DeclaringType` returns implementing `class` so it works as expected.

Reply:
Ok, I change it to `DeclaringType` - I spent some time on this, and actually I think this does not matter and both return the same in this case.

When the `StackTrace` class builds the frames it builds `MethodBase` instances for each frame, where I think there is no real meaning of `ReflectedType` or `DeclaringType` - it's just a stackframe with a MethodTable pointer, it's not a type or an object instance that they have - and they seem to end up with a value that the `DeclaringType` stores in the samples above in both the `ReflectedType` and in the `DeclaringType` as well.

I added some unit tests pinning this behaviour (basically testing this `Base`,  `Derived` example), but the callstack we capture is the same no matter whether we have `DecleringType` or `ReflectedType`. Furthermore I tested it with our sample ASP.NET Core app and there were no `MethodBase` instances in the `StackFrame[] frames` array where `item?.GetMethod()?.ReflectedType?.FullName != item?.GetMethod()?.DeclaringType?.FullName` - they always seem to be the same.

I think this kind of makes sense, because in case of a stacktrace when a method name is resolved we start already from a Method Table pointer, while in case of all the discussions with sample code about `ReflectedType` vs. `DeclaringType` it starts from a type or an instance.

Some interesting code:
[`StackTrace.CaptureStackTrace`](https://github.com/dotnet/coreclr/blob/9773db1e7b1acb3ec75c9cc0e36bd62dcbacd6d5/src/System.Private.CoreLib/src/System/Diagnostics/StackTrace.CoreCLR.cs#L62)

[`StackFrameHelper.GetMethodBase`](https://github.com/dotnet/coreclr/blob/9773db1e7b1acb3ec75c9cc0e36bd62dcbacd6d5/src/System.Private.CoreLib/src/System/Diagnostics/StackFrameHelper.cs#L176)

[`GetTypicalMethodDefinitio`](https://github.com/dotnet/coreclr/blob/a64cb0a41a4ebeb4a61b8b0f1f2eeeff8dd539c9/src/vm/runtimehandles.cpp#L2268)

[`LoadTypicalMethodDefinition `](https://github.com/dotnet/coreclr/blob/3dd303f0004ed4771bc29167df30efda07e4cf7e/src/vm/method.cpp#L1732)